tests/end_to_end/CairoPrecompiles/test_batch_cairo_precompile.py

```diff
@@ -0,0 +1,147 @@
+from typing import Any, List, Tuple
+
+import pytest
+import pytest_asyncio
+from eth_abi import encode
+from eth_utils import keccak
+from hypothesis import given, settings
+from hypothesis import strategies as st
+
+from kakarot_scripts.utils.kakarot import deploy, eth_send_transaction
+from kakarot_scripts.utils.starknet import get_contract, wait_for_transaction
+from tests.utils.errors import cairo_error
+
+EVM_MULTICALLCAIRO_SELECTOR = keccak(
+    text="call_contract(uint256,uint256,uint256,uint256[])"
+)[:4]
+
+
+@pytest_asyncio.fixture(scope="module")
+async def cairo_counter(max_fee, deployer):
+    cairo_counter = get_contract("Counter", provider=deployer)
+
+    yield cairo_counter
+
+    tx = await cairo_counter.functions["set_counter"].invoke_v1(0, max_fee=max_fee)
+    await wait_for_transaction(tx.hash)
+
+
+@pytest_asyncio.fixture(scope="module")
+async def multicall_cairo_counter_caller(owner, cairo_counter):
+    caller_contract = await deploy(
+        "CairoPrecompiles",
+        "MulticallCairoCounterCaller",
+        cairo_counter.address,
+        caller_eoa=owner.starknet_contract,
+    )
+    return caller_contract
+
+
+def encode_starknet_call(call) -> bytes:
```

Comment:
looks like it's only used once so I'd inline it directly in the `prepare_transaction_data`